org/mozilla/jss/crypto/KBKDFDerivedKey.java

```diff
@@ -0,0 +1,107 @@
+package org.mozilla.jss.crypto;
+
+import java.lang.IllegalArgumentException;
+import java.security.InvalidKeyException;
+import java.util.Arrays;
+import java.util.ArrayList;
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+
+import org.mozilla.jss.pkcs11.attrs.CKAttribute;
+import org.mozilla.jss.util.NativeEnclosure;
+
+/**
+ * Class for supporting additional derived keys from PKCS#11 v3.0's
+ * SP800-108 KBKDF implementation.
+ */
+public class KBKDFDerivedKey extends NativeEnclosure {
+    private CKAttribute[] attrs;
+    private long handle;
+
+    public KBKDFDerivedKey() {}
```

Comment:
Is this actually allowed? The `acquireNativeResources()` will fail since `attrs` will be null.

Reply:
You can either call:

```
attrs = new CKAttribute[] { ... };
kdk = new KBKDFDerivedKey(attrs);
```

or:

```
kdk = new KBKDFDerivedKey();
kdk.addAttribute(...);
...
```

And incrementally add attributes for your derived key. `acquireNativeResources(...)` checks to make sure you have, but you have enough rope to make the code look pretty :-) 

Its the same thing with the `KBKDFParameterSpec` class. There's a `validateParameters(...)` method that'll get implemented, along with setters for each of the types... And an empty constructor if you want to build it all up. Else the derived class (`KBKDFCounterParams` / ...) has the constructors which do all of the assignments internally.